File: BabyTracker/Models/Entries.swift
import Foundation
import SwiftData

// MARK: - Feed Entry

@Model
final class FeedEntry {
    @Attribute(.unique) var id: UUID
    var babyID: UUID
    var caregiverID: UUID
    var timestamp: Date
    var notes: String?
    var syncStatus: SyncStatus
    var remoteID: String?
    var createdAt: Date
    var updatedAt: Date

    var feedType: FeedType
    // Breast
    var leftDurationSeconds: Double?
    var rightDurationSeconds: Double?
    var lastSide: BreastSide?
    // Bottle
    var volumeMl: Double?
    var milkType: MilkType?
    var formulaBrand: String?
    // Solids
    var foodName: String?
    var hadReaction: Bool?
    var textureStage: TextureStage?
    // Pump
    var pumpLeftMl: Double?
    var pumpRightMl: Double?
    var pumpStorage: PumpStorage?

    var baby: Baby?

    init(babyID: UUID, caregiverID: UUID, feedType: FeedType, timestamp: Date = Date()) {
        self.id = UUID()
        self.babyID = babyID
        self.caregiverID = caregiverID
        self.timestamp = timestamp
        self.feedType = feedType
        self.syncStatus = .pending
        self.createdAt = Date()
        self.updatedAt = Date()
    }

    var totalDurationSeconds: Double {
        (leftDurationSeconds ?? 0) + (rightDurationSeconds ?? 0)
    }
}

enum FeedType: String, Codable, CaseIterable {
    case breast, bottle, solids, pump
}

enum BreastSide: String, Codable {
    case left, right, both
}

enum MilkType: String, Codable, CaseIterable {
    case breast = "Breast Milk"
    case formula = "Formula"
    case mixed = "Mixed"
}

enum TextureStage: String, Codable, CaseIterable {
    case puree = "Puree"
    case mashed = "Mashed"
    case soft = "Soft Pieces"
    case fingerFood = "Finger Food"
    case table = "Table Food"
}

enum PumpStorage: String, Codable, CaseIterable {
    case fridge = "Refrigerator"
    case freezer = "Freezer"
    case used = "Used Immediately"
}

// MARK: - Sleep Entry

@Model
final class SleepEntry {
    @Attribute(.unique) var id: UUID
    var babyID: UUID
    var caregiverID: UUID
    var timestamp: Date        // start time
    var notes: String?
    var syncStatus: SyncStatus
    var remoteID: String?
    var createdAt: Date
    var updatedAt: Date

    var startTime: Date
    var endTime: Date?
    var sleepType: SleepType
    var location: SleepLocation
    var qualityRating: Int?    // 1-5, optional

    var baby: Baby?

    var durationSeconds: Double? {
        guard let end = endTime else { return nil }
        return end.timeIntervalSince(startTime)
    }

    var isOngoing: Bool { endTime == nil }

    init(babyID: UUID, caregiverID: UUID, startTime: Date = Date(), sleepType: SleepType = .nap) {
        self.id = UUID()
        self.babyID = babyID
        self.caregiverID = caregiverID
        self.timestamp = startTime
        self.startTime = startTime
        self.sleepType = sleepType
        self.location = .crib
        self.syncStatus = .pending
        self.createdAt = Date()
        self.updatedAt = Date()
    }
}

enum SleepType: String, Codable, CaseIterable {
    case nap = "Nap"
    case nighttime = "Nighttime"
}

enum SleepLocation: String, Codable, CaseIterable {
    case crib = "Crib"
    case bassinet = "Bassinet"
    case contactNap = "Contact Nap"
    case stroller = "Stroller"
    case carSeat = "Car Seat"
    case bed = "Bed"
    case other = "Other"
}

// MARK: - Diaper Entry

@Model
final class DiaperEntry {
    @Attribute(.unique) var id: UUID
    var babyID: UUID
    var caregiverID: UUID
    var timestamp: Date
    var notes: String?
    var syncStatus: SyncStatus
    var remoteID: String?
    var createdAt: Date
    var updatedAt: Date

    var diaperType: DiaperType
    var stoolColor: StoolColor?
    var stoolConsistency: StoolConsistency?
    var brand: String?

    var baby: Baby?

    init(babyID: UUID, caregiverID: UUID, diaperType: DiaperType, timestamp: Date = Date()) {
        self.id = UUID()
        self.babyID = babyID
        self.caregiverID = caregiverID
        self.timestamp = timestamp
        self.diaperType = diaperType
        self.syncStatus = .pending
        self.createdAt = Date()
        self.updatedAt = Date()
    }
}

enum DiaperType: String, Codable, CaseIterable {
    case wet = "Wet"
    case dirty = "Dirty"
    case both = "Wet + Dirty"
    case dry = "Dry"
}

enum StoolColor: String, Codable, CaseIterable {
    case yellow = "Yellow"
    case mustard = "Mustard"
    case brown = "Brown"
    case green = "Green"
    case orange = "Orange"
    case red = "Red"       // flag — blood
    case black = "Black"   // flag — meconium or blood
    case white = "White"   // flag — bile duct
}

enum StoolConsistency: String, Codable, CaseIterable {
    case watery = "Watery"
    case loose = "Loose"
    case seedy = "Seedy"
    case pasty = "Pasty"
    case formed = "Formed"
    case hard = "Hard"
}

// MARK: - Temperature Entry

@Model
final class TemperatureEntry {
    @Attribute(.unique) var id: UUID
    var babyID: UUID
    var caregiverID: UUID
    var timestamp: Date
    var notes: String?
    var syncStatus: SyncStatus
    var remoteID: String?
    var createdAt: Date
    var updatedAt: Date

    var valueFahrenheit: Double   // always store as F, convert on display
    var method: TempMethod

    var baby: Baby?

    var valueCelsius: Double { (valueFahrenheit - 32) * 5 / 9 }
    var isFever: Bool { valueFahrenheit >= 100.4 }

    init(babyID: UUID, caregiverID: UUID, valueFahrenheit: Double, method: TempMethod, timestamp: Date = Date()) {
        self.id = UUID()
        self.babyID = babyID
        self.caregiverID = caregiverID
        self.timestamp = timestamp
        self.valueFahrenheit = valueFahrenheit
        self.method = method
        self.syncStatus = .pending
        self.createdAt = Date()
        self.updatedAt = Date()
    }
}

enum TempMethod: String, Codable, CaseIterable {
    case rectal = "Rectal"
    case axillary = "Armpit"
    case temporal = "Temporal"
    case ear = "Ear"
    case oral = "Oral"
}

File: flutter_app/lib/models/entries.dart
import 'sync_types.dart';

// ─── Feed ─────────────────────────────────────────────────────────────────────

enum FeedType { breast, bottle, solids, pump }
enum BreastSide { left, right, both }
enum FormulaType { standard, sensitive, soy, hypoallergenic, other }

class FeedEntry {
  final String id;
  final String babyId;
  final String caregiverId;
  DateTime timestamp;
  FeedType feedType;
  // Breast
  BreastSide? side;
  int? leftDurationSeconds;
  int? rightDurationSeconds;
  // Bottle
  double? amountMl;
  bool? isBreastMilk;
  FormulaType? formulaType;
  // Solids
  String? foodDescription;
  // Pump
  double? pumpedMl;
  String? notes;
  SyncStatus syncStatus;
  String? remoteId;
  final DateTime createdAt;
  DateTime updatedAt;

  FeedEntry({
    required this.id,
    required this.babyId,
    required this.caregiverId,
    required this.timestamp,
    required this.feedType,
    this.side,
    this.leftDurationSeconds,
    this.rightDurationSeconds,
    this.amountMl,
    this.isBreastMilk,
    this.formulaType,
    this.foodDescription,
    this.pumpedMl,
    this.notes,
    this.syncStatus = SyncStatus.pending,
    this.remoteId,
    DateTime? createdAt,
    DateTime? updatedAt,
  })  : createdAt = createdAt ?? DateTime.now(),
        updatedAt = updatedAt ?? DateTime.now();

  int get totalDurationSeconds =>
      (leftDurationSeconds ?? 0) + (rightDurationSeconds ?? 0);

  String get primaryLabel {
    switch (feedType) {
      case FeedType.breast:
        final mins = totalDurationSeconds ~/ 60;
        return '$mins min nursing';
      case FeedType.bottle:
        return amountMl != null ? '${amountMl!.toStringAsFixed(0)} ml' : 'Bottle';
      case FeedType.solids:
        return foodDescription ?? 'Solids';
      case FeedType.pump:
        return pumpedMl != null ? '${pumpedMl!.toStringAsFixed(0)} ml pumped' : 'Pump';
    }
  }

  Map<String, dynamic> toJson() => {
        'id': id,
        'babyId': babyId,
        'caregiverId': caregiverId,
        'timestamp': timestamp.toIso8601String(),
        'feedType': feedType.name,
        'side': side?.name,
        'leftDurationSeconds': leftDurationSeconds,
        'rightDurationSeconds': rightDurationSeconds,
        'amountMl': amountMl,
        'isBreastMilk': isBreastMilk,
        'formulaType': formulaType?.name,
        'foodDescription': foodDescription,
        'pumpedMl': pumpedMl,
        'notes': notes,
        'syncStatus': syncStatus.name,
        'remoteId': remoteId,
        'createdAt': createdAt.toIso8601String(),
        'updatedAt': updatedAt.toIso8601String(),
      };

  factory FeedEntry.fromJson(Map<String, dynamic> j) => FeedEntry(
        id: j['id'] as String,
        babyId: j['babyId'] as String,
        caregiverId: j['caregiverId'] as String,
        timestamp: DateTime.parse(j['timestamp'] as String),
        feedType: FeedType.values.byName(j['feedType'] as String),
        side: j['side'] != null ? BreastSide.values.byName(j['side'] as String) : null,
        leftDurationSeconds: j['leftDurationSeconds'] as int?,
        rightDurationSeconds: j['rightDurationSeconds'] as int?,
        amountMl: (j['amountMl'] as num?)?.toDouble(),
        isBreastMilk: j['isBreastMilk'] as bool?,
        formulaType: j['formulaType'] != null
            ? FormulaType.values.byName(j['formulaType'] as String)
            : null,
        foodDescription: j['foodDescription'] as String?,
        pumpedMl: (j['pumpedMl'] as num?)?.toDouble(),
        notes: j['notes'] as String?,
        syncStatus: SyncStatus.values.byName(j['syncStatus'] as String? ?? 'pending'),
        remoteId: j['remoteId'] as String?,
        createdAt: DateTime.parse(j['createdAt'] as String),
        updatedAt: DateTime.parse(j['updatedAt'] as String),
      );
}

// ─── Sleep ────────────────────────────────────────────────────────────────────

class SleepEntry {
  final String id;
  final String babyId;
  final String caregiverId;
  DateTime timestamp;
  DateTime startTime;
  DateTime? endTime;
  String? notes;
  SyncStatus syncStatus;
  String? remoteId;
  final DateTime createdAt;
  DateTime updatedAt;

  SleepEntry({
    required this.id,
    required this.babyId,
    required this.caregiverId,
    required this.timestamp,
    required this.startTime,
    this.endTime,
    this.notes,
    this.syncStatus = SyncStatus.pending,
    this.remoteId,
    DateTime? createdAt,
    DateTime? updatedAt,
  })  : createdAt = createdAt ?? DateTime.now(),
        updatedAt = updatedAt ?? DateTime.now();

  bool get isOngoing => endTime == null;

  Duration get duration => isOngoing
      ? DateTime.now().difference(startTime)
      : endTime!.difference(startTime);

  String get durationLabel {
    final d = duration;
    final h = d.inHours;
    final m = d.inMinutes % 60;
    if (h > 0) return '${h}h ${m}m';
    return '${m}m';
  }

  Map<String, dynamic> toJson() => {
        'id': id,
        'babyId': babyId,
        'caregiverId': caregiverId,
        'timestamp': timestamp.toIso8601String(),
        'startTime': startTime.toIso8601String(),
        'endTime': endTime?.toIso8601String(),
        'notes': notes,
        'syncStatus': syncStatus.name,
        'remoteId': remoteId,
        'createdAt': createdAt.toIso8601String(),
        'updatedAt': updatedAt.toIso8601String(),
      };

  factory SleepEntry.fromJson(Map<String, dynamic> j) => SleepEntry(
        id: j['id'] as String,
        babyId: j['babyId'] as String,
        caregiverId: j['caregiverId'] as String,
        timestamp: DateTime.parse(j['timestamp'] as String),
        startTime: DateTime.parse(j['startTime'] as String),
        endTime: j['endTime'] != null ? DateTime.parse(j['endTime'] as String) : null,
        notes: j['notes'] as String?,
        syncStatus: SyncStatus.values.byName(j['syncStatus'] as String? ?? 'pending'),
        remoteId: j['remoteId'] as String?,
        createdAt: DateTime.parse(j['createdAt'] as String),
        updatedAt: DateTime.parse(j['updatedAt'] as String),
      );
}

// ─── Diaper ───────────────────────────────────────────────────────────────────

enum DiaperType { wet, dirty, both, dry }
enum StoolColor { yellow, green, brown, orange, red, black, white, other }

class DiaperEntry {
  final String id;
  final String babyId;
  final String caregiverId;
  DateTime timestamp;
  DiaperType diaperType;
  StoolColor? stoolColor;
  bool hasRash;
  String? notes;
  SyncStatus syncStatus;
  String? remoteId;
  final DateTime createdAt;
  DateTime updatedAt;

  DiaperEntry({
    required this.id,
    required this.babyId,
    required this.caregiverId,
    required this.timestamp,
    required this.diaperType,
    this.stoolColor,
    this.hasRash = false,
    this.notes,
    this.syncStatus = SyncStatus.pending,
    this.remoteId,
    DateTime? createdAt,
    DateTime? updatedAt,
  })  : createdAt = createdAt ?? DateTime.now(),
        updatedAt = updatedAt ?? DateTime.now();

  bool get requiresAttention =>
      stoolColor == StoolColor.red ||
      stoolColor == StoolColor.black ||
      stoolColor == StoolColor.white;

  Map<String, dynamic> toJson() => {
        'id': id,
        'babyId': babyId,
        'caregiverId': caregiverId,
        'timestamp': timestamp.toIso8601String(),
        'diaperType': diaperType.name,
        'stoolColor': stoolColor?.name,
        'hasRash': hasRash,
        'notes': notes,
        'syncStatus': syncStatus.name,
        'remoteId': remoteId,
        'createdAt': createdAt.toIso8601String(),
        'updatedAt': updatedAt.toIso8601String(),
      };

  factory DiaperEntry.fromJson(Map<String, dynamic> j) => DiaperEntry(
        id: j['id'] as String,
        babyId: j['babyId'] as String,
        caregiverId: j['caregiverId'] as String,
        timestamp: DateTime.parse(j['timestamp'] as String),
        diaperType: DiaperType.values.byName(j['diaperType'] as String),
        stoolColor: j['stoolColor'] != null
            ? StoolColor.values.byName(j['stoolColor'] as String)
            : null,
        hasRash: j['hasRash'] as bool? ?? false,
        notes: j['notes'] as String?,
        syncStatus: SyncStatus.values.byName(j['syncStatus'] as String? ?? 'pending'),
        remoteId: j['remoteId'] as String?,
        createdAt: DateTime.parse(j['createdAt'] as String),
        updatedAt: DateTime.parse(j['updatedAt'] as String),
      );
}

// ─── Temperature ──────────────────────────────────────────────────────────────

enum TempUnit { fahrenheit, celsius }
enum TempMethod { rectal, axillary, oral, temporal, tympanic }

class TemperatureEntry {
  final String id;
  final String babyId;
  final String caregiverId;
  DateTime timestamp;
  double valueFahrenheit;
  TempMethod method;
  String? notes;
  SyncStatus syncStatus;
  String? remoteId;
  final DateTime createdAt;
  DateTime updatedAt;

  TemperatureEntry({
    required this.id,
    required this.babyId,
    required this.caregiverId,
    required this.timestamp,
    required this.valueFahrenheit,
    this.method = TempMethod.axillary,
    this.notes,
    this.syncStatus = SyncStatus.pending,
    this.remoteId,
    DateTime? createdAt,
    DateTime? updatedAt,
  })  : createdAt = createdAt ?? DateTime.now(),
        updatedAt = updatedAt ?? DateTime.now();

  bool get isFever => valueFahrenheit >= 100.4;

  double get valueCelsius => (valueFahrenheit - 32) * 5 / 9;

  Map<String, dynamic> toJson() => {
        'id': id,
        'babyId': babyId,
        'caregiverId': caregiverId,
        'timestamp': timestamp.toIso8601String(),
        'valueFahrenheit': valueFahrenheit,
        'method': method.name,
        'notes': notes,
        'syncStatus': syncStatus.name,
        'remoteId': remoteId,
        'createdAt': createdAt.toIso8601String(),
        'updatedAt': updatedAt.toIso8601String(),
      };

  factory TemperatureEntry.fromJson(Map<String, dynamic> j) => TemperatureEntry(
        id: j['id'] as String,
        babyId: j['babyId'] as String,
        caregiverId: j['caregiverId'] as String,
        timestamp: DateTime.parse(j['timestamp'] as String),
        valueFahrenheit: (j['valueFahrenheit'] as num).toDouble(),
        method: TempMethod.values.byName(j['method'] as String? ?? 'axillary'),
        notes: j['notes'] as String?,
        syncStatus: SyncStatus.values.byName(j['syncStatus'] as String? ?? 'pending'),
        remoteId: j['remoteId'] as String?,
        createdAt: DateTime.parse(j['createdAt'] as String),
        updatedAt: DateTime.parse(j['updatedAt'] as String),
      );
}
